designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,179 @@
+
+* `project.assets.json`
+* `projectName.ext.nuget.g.props`
+* `projectName.ext.nuget.g.targets`
+
+At build-time these files are loaded and interpreted as necessary.
+This now meant that the NuGet version and build tooling version have an implied compatibility matrix.
+
+### CI experience and compatibility matrices 
+
+The additional tooling options, the combination of project types in one repository now means that there are many ways people run the restore/build gestures on CI.
+
+1. dotnet.exe
+With dotnet.exe, the CI commands would normally look like one of the below options: 
+
+    * dotnet.exe restore & dotnet.exe build --no-restore
+    * dotnet.exe build (restore is implied)
+
+This is the recommended and full proof way if you have only projects that are supported by the [`.NET SDK`](https://github.com/dotnet/sdk).
+packages.config projects *are not* supported with dotnet.exe.
+
+dotnet.exe and it's respective SDK are a self-contained build tool. The restore and build contract is 100% understood and there are no compatibility concerns.
+
+2. MSBuild.exe
+
+With MSBuild.exe, the CI commands would normally look like one of the below options:
+
+* MSBuild.exe /t:restore & MSBuild.exe /t:build
+* MSBuild.exe /restore /t:build
+
+MSBuild.exe has had PackageReference restore support since VS 2017. Starting with VS 2019, update 5, MSBuild now has project specific packages.config support.
+
+Using MSBuild.exe to restore and build the repo means that the restore and build contract is 100% understood.
+
+3. NuGet.exe
+
+NuGet.exe is the oldest tool that can be used to restore on CIs.
+It supports PackageReference and packages.config, but obviously it cannot build, so one still needs to build, usually using MSBuild.exe.
+So the steps are usually:
+
+* NuGet.exe restore
+* MSBuild.exe /t:build
+
+In packages.config, this usually just works because of packages.config's architecture.
+In PackageReference, there's a nuance, because now you have a restore and build steps that are from potential different versions of the tooling.  
+
+Naturally this leads to 2 potential different problems.
+
+#### Old NuGet tooling, newer build tooling - NuGet forward compatibility
+
+In recent years, NuGet restore has added many capabilities to support .NET Core. From FrameworkReference in 5.3 to support for .NET 5 TFMs in 5.8.
+In *most* situations, using an older version of NuGet.exe *will not* cause any problems at build time. 5.8 and .NET 5 in particular are not such versions.
+[NuGet/Home#9756](https://github.com/NuGet/Home/issues/9756) introduces a change that was needed to support .NET 5 tfms and that change in itself required a significant change to how the build tasks interpret the `project.assets.json`.
+An example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/NuGet-Restore-build-issues---projectass/1190427).
+
+#### New NuGet tooling, old build tooling - NuGet backward compatibility

```

Comment:
Is it worthwhile mentioning in either this backwards compat or the previous forwards compat sections, that the "full tfm alias" feature is probably going to need to break both backwards and forwards compat?

maybe this means we can't implement it until VS 17.0/.net 6

Reply:
I'll add it in references.